{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process fastqs from geo into count files \n",
    "NOTE: you will have to the fastqs you wish to process from GEO/SRA and place them in `/data/fastqs/` for this script to function  \n",
    "If you wish to start with processed data, you can skip directly to notebooks in dir `/x02_analysis`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import re\n",
    "\n",
    "from Bio import SeqIO \n",
    "\n",
    "import gzip\n",
    "import os \n",
    "import re \n",
    "import sys\n",
    "import time\n",
    "import bz2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### read in sample info data from geo meta file "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_df = pd.read_excel('../data/meta/GSE139657_meta.xls', \n",
    "                       skiprows=21,nrows=170 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample name</th>\n",
       "      <th>title</th>\n",
       "      <th>source name</th>\n",
       "      <th>organism</th>\n",
       "      <th>characteristics: experiment</th>\n",
       "      <th>molecule</th>\n",
       "      <th>description</th>\n",
       "      <th>processed data file</th>\n",
       "      <th>processed data file .1</th>\n",
       "      <th>raw file</th>\n",
       "      <th>raw file.1</th>\n",
       "      <th>raw file.2</th>\n",
       "      <th>raw file.3</th>\n",
       "      <th>raw file.4</th>\n",
       "      <th>raw file.5</th>\n",
       "      <th>raw file.6</th>\n",
       "      <th>raw file.7</th>\n",
       "      <th>Unnamed: 17</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sample 1</td>\n",
       "      <td>CMV_plasmid_0_p-a_GEN00105092</td>\n",
       "      <td>DNA plasmid</td>\n",
       "      <td>AAV</td>\n",
       "      <td>capsid packaging screen</td>\n",
       "      <td>AAV genome DNA</td>\n",
       "      <td>sequencing of AAV genome to determine packaing...</td>\n",
       "      <td>aav_packaging_all.csv</td>\n",
       "      <td>CMV_plasmid_0_p-a_GEN00105092.csv</td>\n",
       "      <td>LIB030216_GEN00105092_S22_L001_R1.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105092_S22_L004_R2.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105092_S22_L004_R1.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105092_S22_L002_R2.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105092_S22_L002_R1.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105092_S22_L003_R1.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105092_S22_L003_R2.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105092_S22_L001_R2.fastq.bz2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sample 2</td>\n",
       "      <td>CMV_plasmid_0_p-b_GEN00113847</td>\n",
       "      <td>DNA plasmid</td>\n",
       "      <td>AAV</td>\n",
       "      <td>capsid packaging screen</td>\n",
       "      <td>AAV genome DNA</td>\n",
       "      <td>sequencing of AAV genome to determine packaing...</td>\n",
       "      <td>aav_packaging_all.csv</td>\n",
       "      <td>CMV_plasmid_0_p-b_GEN00113847.csv</td>\n",
       "      <td>LIB031744_GEN00113847_S101_L004_R1.fastq.bz2</td>\n",
       "      <td>LIB031744_GEN00113847_S101_L003_R1.fastq.bz2</td>\n",
       "      <td>LIB031744_GEN00113847_S101_L003_R2.fastq.bz2</td>\n",
       "      <td>LIB031744_GEN00113847_S101_L004_R2.fastq.bz2</td>\n",
       "      <td>LIB031744_GEN00113847_S101_L002_R1.fastq.bz2</td>\n",
       "      <td>LIB031744_GEN00113847_S101_L002_R2.fastq.bz2</td>\n",
       "      <td>LIB031744_GEN00113847_S101_L001_R2.fastq.bz2</td>\n",
       "      <td>LIB031744_GEN00113847_S101_L001_R1.fastq.bz2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sample 3</td>\n",
       "      <td>CMV_plasmid_0_p-c_GEN00113848</td>\n",
       "      <td>DNA plasmid</td>\n",
       "      <td>AAV</td>\n",
       "      <td>capsid packaging screen</td>\n",
       "      <td>AAV genome DNA</td>\n",
       "      <td>sequencing of AAV genome to determine packaing...</td>\n",
       "      <td>aav_packaging_all.csv</td>\n",
       "      <td>CMV_plasmid_0_p-c_GEN00113848.csv</td>\n",
       "      <td>LIB031744_GEN00113848_S102_L004_R1.fastq.bz2</td>\n",
       "      <td>LIB031744_GEN00113848_S102_L001_R2.fastq.bz2</td>\n",
       "      <td>LIB031744_GEN00113848_S102_L003_R2.fastq.bz2</td>\n",
       "      <td>LIB031744_GEN00113848_S102_L004_R2.fastq.bz2</td>\n",
       "      <td>LIB031744_GEN00113848_S102_L002_R1.fastq.bz2</td>\n",
       "      <td>LIB031744_GEN00113848_S102_L001_R1.fastq.bz2</td>\n",
       "      <td>LIB031744_GEN00113848_S102_L002_R2.fastq.bz2</td>\n",
       "      <td>LIB031744_GEN00113848_S102_L003_R1.fastq.bz2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sample 4</td>\n",
       "      <td>CMV_virus_1_a_GEN00105095</td>\n",
       "      <td>AAV Capsid</td>\n",
       "      <td>AAV</td>\n",
       "      <td>capsid packaging screen</td>\n",
       "      <td>AAV genome DNA</td>\n",
       "      <td>sequencing of AAV genome to determine packaing...</td>\n",
       "      <td>aav_packaging_all.csv</td>\n",
       "      <td>CMV_virus_1_a_GEN00105095.csv</td>\n",
       "      <td>LIB030216_GEN00105095_S25_L001_R2.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105095_S25_L003_R2.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105095_S25_L003_R1.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105095_S25_L002_R2.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105095_S25_L004_R1.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105095_S25_L001_R1.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105095_S25_L004_R2.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105095_S25_L002_R1.fastq.bz2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sample 5</td>\n",
       "      <td>CMV_virus_1_a_GEN00105099</td>\n",
       "      <td>AAV Capsid</td>\n",
       "      <td>AAV</td>\n",
       "      <td>capsid packaging screen</td>\n",
       "      <td>AAV genome DNA</td>\n",
       "      <td>sequencing of AAV genome to determine packaing...</td>\n",
       "      <td>aav_packaging_all.csv</td>\n",
       "      <td>CMV_virus_1_a_GEN00105099.csv</td>\n",
       "      <td>LIB030216_GEN00105099_S29_L003_R1.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105099_S29_L002_R1.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105099_S29_L001_R2.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105099_S29_L004_R2.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105099_S29_L001_R1.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105099_S29_L002_R2.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105099_S29_L003_R2.fastq.bz2</td>\n",
       "      <td>LIB030216_GEN00105099_S29_L004_R1.fastq.bz2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample name                          title  source name organism  \\\n",
       "0    Sample 1  CMV_plasmid_0_p-a_GEN00105092  DNA plasmid      AAV   \n",
       "1    Sample 2  CMV_plasmid_0_p-b_GEN00113847  DNA plasmid      AAV   \n",
       "2    Sample 3  CMV_plasmid_0_p-c_GEN00113848  DNA plasmid      AAV   \n",
       "3    Sample 4      CMV_virus_1_a_GEN00105095   AAV Capsid      AAV   \n",
       "4    Sample 5      CMV_virus_1_a_GEN00105099   AAV Capsid      AAV   \n",
       "\n",
       "  characteristics: experiment        molecule  \\\n",
       "0     capsid packaging screen  AAV genome DNA   \n",
       "1     capsid packaging screen  AAV genome DNA   \n",
       "2     capsid packaging screen  AAV genome DNA   \n",
       "3     capsid packaging screen  AAV genome DNA   \n",
       "4     capsid packaging screen  AAV genome DNA   \n",
       "\n",
       "                                         description   processed data file   \\\n",
       "0  sequencing of AAV genome to determine packaing...  aav_packaging_all.csv   \n",
       "1  sequencing of AAV genome to determine packaing...  aav_packaging_all.csv   \n",
       "2  sequencing of AAV genome to determine packaing...  aav_packaging_all.csv   \n",
       "3  sequencing of AAV genome to determine packaing...  aav_packaging_all.csv   \n",
       "4  sequencing of AAV genome to determine packaing...  aav_packaging_all.csv   \n",
       "\n",
       "              processed data file .1  \\\n",
       "0  CMV_plasmid_0_p-a_GEN00105092.csv   \n",
       "1  CMV_plasmid_0_p-b_GEN00113847.csv   \n",
       "2  CMV_plasmid_0_p-c_GEN00113848.csv   \n",
       "3      CMV_virus_1_a_GEN00105095.csv   \n",
       "4      CMV_virus_1_a_GEN00105099.csv   \n",
       "\n",
       "                                       raw file  \\\n",
       "0   LIB030216_GEN00105092_S22_L001_R1.fastq.bz2   \n",
       "1  LIB031744_GEN00113847_S101_L004_R1.fastq.bz2   \n",
       "2  LIB031744_GEN00113848_S102_L004_R1.fastq.bz2   \n",
       "3   LIB030216_GEN00105095_S25_L001_R2.fastq.bz2   \n",
       "4   LIB030216_GEN00105099_S29_L003_R1.fastq.bz2   \n",
       "\n",
       "                                     raw file.1  \\\n",
       "0   LIB030216_GEN00105092_S22_L004_R2.fastq.bz2   \n",
       "1  LIB031744_GEN00113847_S101_L003_R1.fastq.bz2   \n",
       "2  LIB031744_GEN00113848_S102_L001_R2.fastq.bz2   \n",
       "3   LIB030216_GEN00105095_S25_L003_R2.fastq.bz2   \n",
       "4   LIB030216_GEN00105099_S29_L002_R1.fastq.bz2   \n",
       "\n",
       "                                     raw file.2  \\\n",
       "0   LIB030216_GEN00105092_S22_L004_R1.fastq.bz2   \n",
       "1  LIB031744_GEN00113847_S101_L003_R2.fastq.bz2   \n",
       "2  LIB031744_GEN00113848_S102_L003_R2.fastq.bz2   \n",
       "3   LIB030216_GEN00105095_S25_L003_R1.fastq.bz2   \n",
       "4   LIB030216_GEN00105099_S29_L001_R2.fastq.bz2   \n",
       "\n",
       "                                     raw file.3  \\\n",
       "0   LIB030216_GEN00105092_S22_L002_R2.fastq.bz2   \n",
       "1  LIB031744_GEN00113847_S101_L004_R2.fastq.bz2   \n",
       "2  LIB031744_GEN00113848_S102_L004_R2.fastq.bz2   \n",
       "3   LIB030216_GEN00105095_S25_L002_R2.fastq.bz2   \n",
       "4   LIB030216_GEN00105099_S29_L004_R2.fastq.bz2   \n",
       "\n",
       "                                     raw file.4  \\\n",
       "0   LIB030216_GEN00105092_S22_L002_R1.fastq.bz2   \n",
       "1  LIB031744_GEN00113847_S101_L002_R1.fastq.bz2   \n",
       "2  LIB031744_GEN00113848_S102_L002_R1.fastq.bz2   \n",
       "3   LIB030216_GEN00105095_S25_L004_R1.fastq.bz2   \n",
       "4   LIB030216_GEN00105099_S29_L001_R1.fastq.bz2   \n",
       "\n",
       "                                     raw file.5  \\\n",
       "0   LIB030216_GEN00105092_S22_L003_R1.fastq.bz2   \n",
       "1  LIB031744_GEN00113847_S101_L002_R2.fastq.bz2   \n",
       "2  LIB031744_GEN00113848_S102_L001_R1.fastq.bz2   \n",
       "3   LIB030216_GEN00105095_S25_L001_R1.fastq.bz2   \n",
       "4   LIB030216_GEN00105099_S29_L002_R2.fastq.bz2   \n",
       "\n",
       "                                     raw file.6  \\\n",
       "0   LIB030216_GEN00105092_S22_L003_R2.fastq.bz2   \n",
       "1  LIB031744_GEN00113847_S101_L001_R2.fastq.bz2   \n",
       "2  LIB031744_GEN00113848_S102_L002_R2.fastq.bz2   \n",
       "3   LIB030216_GEN00105095_S25_L004_R2.fastq.bz2   \n",
       "4   LIB030216_GEN00105099_S29_L003_R2.fastq.bz2   \n",
       "\n",
       "                                     raw file.7 Unnamed: 17  \n",
       "0   LIB030216_GEN00105092_S22_L001_R2.fastq.bz2         NaN  \n",
       "1  LIB031744_GEN00113847_S101_L001_R1.fastq.bz2         NaN  \n",
       "2  LIB031744_GEN00113848_S102_L003_R1.fastq.bz2         NaN  \n",
       "3   LIB030216_GEN00105095_S25_L002_R1.fastq.bz2         NaN  \n",
       "4   LIB030216_GEN00105099_S29_L004_R1.fastq.bz2         NaN  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "functions used to count barcodes and write counts "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tabulate(key, counts):\n",
    "    # add 1 to count for this key, or add it to the dictionary\n",
    "    if key in counts:\n",
    "        counts[key] += 1\n",
    "    else:\n",
    "        counts[key] = 1\n",
    "\n",
    "def write_counts(dict, outputfile):\n",
    "    # write sequences and counts to a text file\n",
    "#     print outputfile\n",
    "    file_out = open(outputfile, 'w')\n",
    "    file_out.write('barcode,count\\n')\n",
    "    for w in sorted(dict, key=dict.get, reverse=True):\n",
    "        file_out.write('{seq},{num}\\n'.format(seq=w, num=str(dict[w])))\n",
    "    file_out.close() \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### using a row from the geo meta data, compute the counts for each barcode across all associated fastqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_ligated_bars(output_path = '../data/',\n",
    "                       count_dir_name = 'counts',\n",
    "                       log_out_name = 'count_log_file',\n",
    "                       meta_df_row = None, \n",
    "                      limit=None):\n",
    "    \n",
    "    lib_file = os.path.join(output_path,'fastq',meta_df_row['raw file'])\n",
    "    save_name = meta_df_row['title']\n",
    "#     print (gen_id)\n",
    "    output_counts_dir = os.path.join(output_path, count_dir_name)\n",
    "    out_file_full = os.path.join(output_counts_dir , '%s.txt' % log_out_name)\n",
    "#     if from_top:\n",
    "#         if os.path.isfile(out_file_full):\n",
    "#             os.remove(out_file_full)\n",
    "    if os.path.isfile(out_file_full):\n",
    "        log = open(out_file_full, 'a')\n",
    "    else:\n",
    "        log = open(out_file_full, 'w')\n",
    "        log.write('reads\\tbarcodes\\tbarcodes_per_read\\tlib_name\\n')\n",
    "    lig_dict ={}\n",
    "    bar_count = 0\n",
    "    read_count = 0 \n",
    "#     print 'counting lib %s' % lib_file\n",
    "    lig_dict = {}\n",
    " \n",
    "    full_file_list = []\n",
    "    for lane in ['L001', 'L002', 'L003','L004']:\n",
    "        for read_direction in ['R1','R2']:\n",
    "            lane_sub = re.sub('L001', lane, lib_file)\n",
    "            full_file_name = re.sub('R1', read_direction,lane_sub )\n",
    "            print (full_file_name)\n",
    "            if os.path.isfile(full_file_name) == False:\n",
    "                print (\"file %s does not exsits\" % full_file_name)\n",
    "                continue\n",
    "            full_file_list.append(full_file_name)\n",
    "            if 'gz' in full_file_name:\n",
    "                handle = gzip.open(full_file_name)\n",
    "            if 'bz' in full_file_name:\n",
    "                print (\"bz file\")\n",
    "                handle = bz2.open(full_file_name, 'rt')\n",
    "\n",
    "            for idx, read in enumerate(SeqIO.parse(handle,'fastq')):\n",
    "                read_count += 1 \n",
    "                if limit:\n",
    "                    if idx > limit : break \n",
    "                if 'R1' in read_direction:\n",
    "                    seq = read.seq\n",
    "                else:\n",
    "                    seq = read.seq.reverse_complement()\n",
    "    #             print seq # to debug\n",
    "                middle_expr = '(?<=CCAC)([ACTG]{20})(?=CCAC)'\n",
    "                matches = re.findall(middle_expr, str(seq))\n",
    "                if matches:\n",
    "                    for match in matches:\n",
    "                        bar_count += 1\n",
    "                        tabulate(match, lig_dict)\n",
    "\n",
    "    print (\"writing counts for %s\" % save_name)\n",
    "    assert len(full_file_list) == 8 , full_file_list\n",
    "    log.write('%s\\t%s\\t%s\\t%s\\n' % (read_count,bar_count, (bar_count/float(read_count)), save_name))\n",
    "    write_counts(lig_dict, os.path.join(output_counts_dir, '%s.csv' % (save_name)))\n",
    "    log.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "example counting the barcodes for one fo the sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../data/fastq/LIB031744_GEN00113848_S102_L004_R1.fastq.bz2\n",
      "bz file\n",
      "../data/fastq/LIB031744_GEN00113848_S102_L004_R2.fastq.bz2\n",
      "bz file\n",
      "../data/fastq/LIB031744_GEN00113848_S102_L004_R1.fastq.bz2\n",
      "bz file\n",
      "../data/fastq/LIB031744_GEN00113848_S102_L004_R2.fastq.bz2\n",
      "bz file\n",
      "../data/fastq/LIB031744_GEN00113848_S102_L004_R1.fastq.bz2\n",
      "bz file\n",
      "../data/fastq/LIB031744_GEN00113848_S102_L004_R2.fastq.bz2\n",
      "bz file\n",
      "../data/fastq/LIB031744_GEN00113848_S102_L004_R1.fastq.bz2\n",
      "bz file\n",
      "../data/fastq/LIB031744_GEN00113848_S102_L004_R2.fastq.bz2\n",
      "bz file\n",
      "writing counts for CMV_plasmid_0_p-c_GEN00113848\n"
     ]
    }
   ],
   "source": [
    "count_ligated_bars(meta_df_row = meta_df.iloc[2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### merge the barcode counts onto the meta-data table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "barcode_to_mutant_table = pd.read_csv('../data/meta/AAV2scan_chip_lookup_table.txt')\n",
    "def assemble_dataframe(meta_df_row=meta_df.iloc[2], mutant_info_df = barcode_to_mutant_table ,save=False):\n",
    "    filename = meta_df_row['title']\n",
    "    count_file = os.path.join('../data','counts','%s.csv' % filename )\n",
    "    count_df = pd.read_csv(count_file)\n",
    "    count_with_mutant_info_df = barcode_to_mutant_table.merge(count_df, on='barcode',how='left')\n",
    "    if save:\n",
    "        count_with_mutant_info_df.to_csv(os.path.join('../data/counts', \"%s_dataframe.csv\" % filename))\n",
    "    return count_with_mutant_info_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "example for counts above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aa</th>\n",
       "      <th>abs_pos</th>\n",
       "      <th>barcode</th>\n",
       "      <th>codon</th>\n",
       "      <th>enzyme</th>\n",
       "      <th>is_wt_aa</th>\n",
       "      <th>is_wt_codon</th>\n",
       "      <th>lib_type</th>\n",
       "      <th>tile_num</th>\n",
       "      <th>wt_bc</th>\n",
       "      <th>aa-codon</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>*</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CACTGTCACACACTGACACT</td>\n",
       "      <td>TAA</td>\n",
       "      <td>bbsi</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>sub</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>*-TAA</td>\n",
       "      <td>632.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>*</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CTGTGAGTGTGAGAGACACT</td>\n",
       "      <td>TAA</td>\n",
       "      <td>bbsi</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>sub</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>*-TAA</td>\n",
       "      <td>616.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>*</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CTCTCACACAGTGAGTCTGA</td>\n",
       "      <td>TAG</td>\n",
       "      <td>bbsi</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>sub</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>*-TAG</td>\n",
       "      <td>108.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>*</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CAGAGACAGAGTCTGTCACT</td>\n",
       "      <td>TAG</td>\n",
       "      <td>bbsi</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>sub</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>*-TAG</td>\n",
       "      <td>136.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>*</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ACACTGTCTCTGTCAGACAG</td>\n",
       "      <td>TGA</td>\n",
       "      <td>bbsi</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>sub</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>*-TGA</td>\n",
       "      <td>1048.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  aa  abs_pos               barcode codon enzyme  is_wt_aa  is_wt_codon  \\\n",
       "0  *      1.0  CACTGTCACACACTGACACT   TAA   bbsi         0            0   \n",
       "1  *      1.0  CTGTGAGTGTGAGAGACACT   TAA   bbsi         0            0   \n",
       "2  *      1.0  CTCTCACACAGTGAGTCTGA   TAG   bbsi         0            0   \n",
       "3  *      1.0  CAGAGACAGAGTCTGTCACT   TAG   bbsi         0            0   \n",
       "4  *      1.0  ACACTGTCTCTGTCAGACAG   TGA   bbsi         0            0   \n",
       "\n",
       "  lib_type  tile_num  wt_bc aa-codon   count  \n",
       "0      sub       0.0      0    *-TAA   632.0  \n",
       "1      sub       0.0      0    *-TAA   616.0  \n",
       "2      sub       0.0      0    *-TAG   108.0  \n",
       "3      sub       0.0      0    *-TAG   136.0  \n",
       "4      sub       0.0      0    *-TGA  1048.0  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_data_with_mutant_info_df = assemble_dataframe()\n",
    "count_data_with_mutant_info_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "aav_venv_v2",
   "language": "python",
   "name": "aav_venv_v2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
